rsmtool/rsmpredict.py

```diff
@@ -25,6 +28,165 @@
+       re-scale the predictions. May be read from the "postprocessing_params.csv"
+       file under the "output" directory of the RSMTool experiment used to train
+       the model. If ``None``, predictions will not be scaled.
+       Defaults to ``None``.
+    logger : logging object, optional
+        A logging object. If ``None`` is passed, get logger from ``__name__``.
+        Defaults to ``None``.
+
+    Returns
+    -------
+    dict[str, float]
+        A dictionary containing the raw, scaled, trimmed, and rounded
+        predictions for the input features. It always contains the
+        "raw" key and may contain the following additional keys depending
+        on the availability of the various optional arguments:
+        "raw_trim", "raw_trim_round", "scale", "scale_trim",
+        and "scale_trim_round".
+
+    Raises
+    ------
+    ValueError
+        If ``input_features`` contains any non-numeric features.
+    """
+    # initialize a logger if none provided
+    logger = logger if logger else logging.getLogger(__name__)
+
+    # instantiate a feature preprocessor
+    preprocessor = FeaturePreprocessor(logger=logger)
+
+    # convert the given features to a data frame and add the "spkitemid" column
+    df_input_features = pd.DataFrame([input_features])
+    df_input_features["spkitemid"] = "RESPONSE"
+
+    # preprocess the input features so that they match what the model expects
+    try:
+        df_processed_features, _ = preprocessor.preprocess_new_data(
+            df_input_features, df_feature_info
+        )
+    except ValueError:
+        raise ValueError("Input features must not contain non-numeric values.") from None
+
+    # now compute the raw prediction for the given features
+    df_predictions = modeler.predict(df_processed_features)
+
+    # trim raw predictions if ``trim_min`` and ``trim_max`` were specified
+    if trim_min and trim_max:
+        df_predictions["raw_trim"] = preprocessor.trim(
+            df_predictions["raw"], trim_min, trim_max, trim_tolerance
+        )
+        df_predictions["raw_trim_round"] = np.rint(df_predictions['raw_trim']).astype("int64")
+
+    # compute scaled predictions if the scaling parameters are available
+    if train_predictions_mean and train_predictions_sd and h1_mean and h1_sd:
+        df_predictions["scale"] = (
+            (df_predictions["raw"] - train_predictions_mean) / train_predictions_sd
+        ) * h1_sd + h1_mean
+
+        # apply trimming to scaled scores too, if possible
+        if trim_min and trim_max:
+            df_predictions["scale_trim"] = preprocessor.trim(
+                df_predictions["scale"], trim_min, trim_max, trim_tolerance
+            )
+            df_predictions["scale_trim_round"] = np.rint(df_predictions['scale_trim']).astype("int64")
+
+    # return the predictions as a dictionary but without the ID column
+    return df_predictions.drop("spkitemid", axis="columns").to_dict(orient="records")[0]
```

Comment:
Without the ID column because there’s only one sample being processed at a time, right?

Reply:
It's because the ID was just a dummy one we added – it's not actually part of the input.

Reply:
Is that dummy ID required to use preprocessor.trim? I'm wondering if it would be possible to make that drop earlier in the code at like L171 so that we don't have to add the second drop in the loop at L174

Reply:
Ah, good call, `trim()` just needs the values and not the data frame itself. Fixed.